core/src/main/java/io/questdb/cairo/TableUtils.java

```diff
@@ -693,13 +695,53 @@ public static void txnPartitionConditionally(CharSink path, long txn) {
         }
     }
 
+    public static void safeReadTxn(TxReader txReader, MicrosecondClock microsecondClock, long spinLockTimeoutUs) {
+        long txn;
+        int loadCount = 0;
+        long deadline = microsecondClock.getTicks() + spinLockTimeoutUs;
+        while (true) {
+            txn = txReader.unsafeReadTxnCheck();
+            Unsafe.getUnsafe().loadFence();
+
+            int symbolColumnCount = txReader.unsafeReadSymbolColumnCount();
+            int partitionSegmentSize = txReader.unsafeReadPartitionSegmentSize(symbolColumnCount);
+            Unsafe.getUnsafe().loadFence();
+
+            if (txn == txReader.unsafeReadTxn()) {
+                txReader.unsafeLoadAll(symbolColumnCount, partitionSegmentSize, loadCount++ > 0);
+                Unsafe.getUnsafe().loadFence();
+
+                if (txn == txReader.unsafeReadTxn()) {
+                    // All good, snapshot read
+                    return;
+                }
+            }
+            // This is unlucky, sequences have changed while we were reading transaction data
+            // We must discard and try again
+            if (microsecondClock.getTicks() > deadline) {
+                LOG.error().$("tx read timeout [timeout=").$(spinLockTimeoutUs).utf8("μs]").$();
+                throw CairoException.instance(0).put("Transaction read timeout");
+            }
+            Os.pause();
+        }
+    }
+
+    public static void unsafeReadTxFile(TxReader txReader) {
+        int symbolColumnCount = txReader.unsafeReadSymbolColumnCount();
+        int partitionSegmentSize = txReader.unsafeReadPartitionSegmentSize(symbolColumnCount);
+        txReader.unsafeLoadAll(symbolColumnCount, partitionSegmentSize, true);
+    }
+
     public static void validate(
-            FilesFacade ff,
             MemoryMR metaMem,
             LowerCaseCharSequenceIntHashMap nameIndex,
             int expectedVersion
     ) {
         try {
+            long fileLength = metaMem.size();
```

Comment:
`fileLength` could be confusing. The returned value is not guaranteed to be exact file length. Perhaps call variable `memSize`?

Reply:
done in ed43dd5